File: linda-core/src/main/java/it/unibo/coordination/utils/events/AbstractEventSourceImpl.java
package it.unibo.coordination.utils.events;

import java.util.Collection;
import java.util.concurrent.CompletableFuture;

abstract class AbstractEventSourceImpl<T> implements EventSource<T>, SyncEventEmitter<T>, AsyncEventEmitter<T> {

    protected abstract Collection<EventListener<T>> getEventListeners();

    @Override
    public void bind(EventListener<T> eventListener) {
        getEventListeners().add(eventListener);
    }

    @Override
    public void unbind(EventListener<T> eventListener) {
        getEventListeners().remove(eventListener);
    }

    @Override
    public void unbindAll() {
        getEventListeners().clear();
    }

    @Override
    public abstract T syncEmit(T data);

    @Override
    public abstract CompletableFuture<T> asyncEmit(T event);

    @Override
    public EventSource<T> getEventSource() {
        return this;
    }
}

File: utils/src/main/kotlin/it/unibo/coordination/utils/events/AbstractEventSourceImpl.kt
package it.unibo.coordination.utils.events

import java.util.concurrent.CompletableFuture

internal abstract class AbstractEventSourceImpl<T> : EventSource<T>, SyncEventEmitter<T>, AsyncEventEmitter<T> {

    protected abstract val eventListeners: MutableCollection<EventListener<T>>

    override fun bind(listener: EventListener<T>) {
        eventListeners.add(listener)
    }

    override fun unbind(listener: EventListener<T>) {
        eventListeners.remove(listener)
    }

    override fun unbindAll() {
        eventListeners.clear()
    }

    abstract override fun syncEmit(data: T): T

    abstract override fun asyncEmit(event: T): CompletableFuture<T>

    override val eventSource: EventSource<T>
        get() = this
}